File: BFS/Minimum Depth of Binary Tree/Minimum Depth of Binary Tree.java

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int minDepth(TreeNode root) {
        if(root == null) return 0;
        Queue<TreeNode> q = new LinkedList<>();
        int level = 0;
        q.offer(root);
        while(!q.isEmpty()){
            int size = q.size();
            level++;
            for(int i = 0; i < size; i++){
                if(q.peek().left != null) q.offer(q.peek().left);
                 if(q.peek().right != null) q.offer(q.peek().right);
                 if(q.peek().left == null && q.peek().right == null) 
                    return level;
                q.poll();
            }
        }
        return level;
    }
}

File: BFS/Minimum Depth of Binary Tree/Minimum Depth of Binary Tree.cpp

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    queue<TreeNode*> q;
    int BFS(TreeNode* root){
        int level = 0;
        q.push(root);
        while(!q.empty()){
            int n = q.size();
            level++;
            while(n--){
                TreeNode* temp = q.front();
                q.pop();
                if(temp->left ) q.push(temp->left);
                if(temp->right) q.push(temp->right);
              // to return shortest leaf that has no childern 
                if(temp->left == NULL && temp->right == NULL) return level;
            }
        }
        return level;
    }
    int minDepth(TreeNode* root) {
        if(root == NULL) return 0;
        return BFS(root);
    }
};
